File: ios/TimelineView.m
//
//  TimelineView.m
//  Timeline
//
//  Created by Admin on 01.10.16.
//  Copyright © 2016 savelichalex. All rights reserved.
//

#import <math.h>
#import "TimelineView.h"

@implementation TimelineView

- (void)setTPosition:(float)value {
    _tPosition = value;
    
    [self setNeedsDisplay];
}

- (void)setCountWordsOnScreen:(int)countWordsOnScreen {
    if (countWordsOnScreen % 2 == 0) {
        _countWordsOnScreen = countWordsOnScreen + 1;
    } else {
        _countWordsOnScreen = countWordsOnScreen;
    }
    [self setNeedsDisplay];
}

- (void)setTimestamps:(NSArray *)timestamps {
    _timestamps = timestamps;
    self->countOfLines = (int)[timestamps count];
    
    [self setNeedsDisplay];
}

- (void)setMinAlpha:(float)minAlpha {
    _minAlpha = minAlpha;
    
    [self setNeedsDisplay];
}

- (void)setMinWidthRatio:(float)minWidthRatio {
    _minWidthRatio = minWidthRatio;
    
    [self setNeedsDisplay];
}

- (void)setLineColor:(UIColor *)lineColor {
  _lineColor = lineColor;
  
  [self setNeedsDisplay];
}

- (void)drawRect:(CGRect)rect {
//    [[UIColor whiteColor] setFill];
//    UIRectFill(self.bounds);
    CGFloat height = self.bounds.size.height;
    self->offsetBetweenLines = height / self->countOfLines;
    self->activeLine = (CGFloat) self.tPosition / self->offsetBetweenLines;
    self->activeLineInt = (int)self->activeLine;
    self->n = self.countWordsOnScreen / 2; // TODO: think about better name for `n`
    self->nRatio = 0.55 + (float)n * 0.05;
    for (int i = 0; i < self->countOfLines; i++) {
        [self drawLine:i];
    }
}

- (void)drawLine:(int)lineNumber {
    if ([self isMainLine:lineNumber] == YES) {
        [self drawMainLine:lineNumber];
    } else {
        [self drawOtherLine:lineNumber];
    }
}

- (BOOL)isMainLine:(int)currentLine {
    float fraction = [self getActiveFraction:self->activeLine];
    int minActiveLine = self->activeLineInt;
    int maxActiveLine = minActiveLine + 1;
    if(fraction < 0.5 && currentLine == minActiveLine) {
        return YES;
    } else if (fraction > 0.5 && currentLine == maxActiveLine){
        return YES;
    }
    return NO;
}

- (void)drawMainLine:(int)lineNumber {
    CGFloat up = lineNumber * self->offsetBetweenLines;
    UIBezierPath *line = [[UIBezierPath alloc] init];
    CGFloat width = [self getLineWidth:lineNumber];
    CGFloat fullWidth = self.bounds.size.width;
    line.lineWidth = 2;
    [self.lineColor setStroke];
    
    NSMutableParagraphStyle *pStyle = [[NSMutableParagraphStyle alloc] init];
    pStyle.alignment = NSTextAlignmentCenter;
    
    UIFont *mainLineFont = [UIFont preferredFontForTextStyle:UIFontTextStyleBody];
    mainLineFont = [mainLineFont fontWithSize:self->offsetBetweenLines];
    
    NSAttributedString *mainLineString = [[NSAttributedString alloc] initWithString:[NSString stringWithFormat:@"%f", self.tPosition] attributes:@{ NSFontAttributeName: mainLineFont, NSParagraphStyleAttributeName: pStyle, NSForegroundColorAttributeName: self.lineColor}];
    
    CGRect mainLineBounds;
    mainLineBounds.size = CGSizeMake(mainLineString.size.width*1.1, mainLineString.size.height);
    mainLineBounds.origin = CGPointMake(width / 2 - mainLineBounds.size.width / 2 + (fullWidth - width), up - mainLineBounds.size.height / 2);
    UIRectFill(mainLineBounds);
    [mainLineString drawInRect:mainLineBounds];
  
    // draw main line with 2 lines
    [line moveToPoint:CGPointMake(fullWidth, up)];
    [line addLineToPoint:CGPointMake(mainLineBounds.origin.x + mainLineBounds.size.width, up)];
    [line moveToPoint:CGPointMake(mainLineBounds.origin.x, up)];
    [line addLineToPoint:CGPointMake(fullWidth - width, up)];
    [line stroke];
}

- (void)drawOtherLine:(int)lineNumber {
    CGFloat up = lineNumber * self->offsetBetweenLines;
    UIBezierPath *line = [[UIBezierPath alloc] init];
    CGFloat width = [self getLineWidth:lineNumber];
    CGFloat fullWidth = self.bounds.size.width;
    CGPoint startPoint = CGPointMake(fullWidth, up);
    CGPoint endPoint = CGPointMake(fullWidth - width, up);
    [line moveToPoint:startPoint];
    [line addLineToPoint:endPoint];
    line.lineWidth = 0.5;
  
    CGFloat red;
    CGFloat green;
    CGFloat blue;
    CGFloat alpha;
    [self.lineColor getRed:&red green:&green blue:&blue alpha:&alpha];
    [[UIColor colorWithRed:red green:green blue:blue alpha:[self getLineAlpha:lineNumber]] setStroke];
    [line stroke];
}

- (float)getLineAlpha:(int)currentLine {
    return self.minAlpha + (1 - ((float)abs(currentLine - self->activeLineInt)/(float)self->countOfLines) * (1 - self.minAlpha));
}

- (CGFloat)getLineWidth:(int)currentLine {
    float additionalFactor = [self getAdditionalWidthRatio:currentLine];
    float betweenBiggestFactor = additionalFactor * [self getBetweenBiggestWidthRatio:currentLine];
    return (self.minWidthRatio +
            additionalFactor * (self->nRatio - self.minWidthRatio) +
            betweenBiggestFactor * (1 - self->nRatio)) * self.bounds.size.width;
}

- (float)getAdditionalWidthRatio:(int)currentLine {
    int linesToActive = [self getCountOfLinesToActive:currentLine];
    float linesRatio = 8.0 - self->n;
    if (linesRatio < 0.0) {
        linesRatio = 3.0;
    }
    return (1.0 - tanhf(linesToActive * linesRatio / 10));
}

- (float)getBetweenBiggestWidthRatio:(int)currentLine {
    float fraction = [self getActiveFraction:self->activeLine];
    if (currentLine <= self->activeLineInt || fraction == 0.0) {
        return (1.0 - fraction);
    } else {
        return fraction;
    }
}

- (int)getCountOfLinesToActive:(int)currentLine {
    float fraction = [self getActiveFraction:self->activeLine];
    int minActiveLine = self->activeLineInt;
    int maxActiveLine = minActiveLine + 1;
    if (currentLine <= self->activeLineInt || fraction == 0.0) {
        return abs(currentLine - minActiveLine);
    } else {
        return abs(currentLine - maxActiveLine);
    }
}

- (float)getActiveFraction:(CGFloat)aLine {
    return aLine - (int)aLine;
}

- (void)initHelper {
    self.tPosition = 0.0;
    self->countOfLines = 0;
    self.countWordsOnScreen = 11;
    self.minAlpha = 0.1;
    self.minWidthRatio = 0.3;
    self.lineColor = [UIColor blackColor];
}

- (id)initWithFrame:(CGRect)rect {
    if((self = [super initWithFrame:rect])) {
        [self initHelper];
    }
    return self;
}

- (id)init {
    if ((self = [super init])) {
        [self initHelper];
    }
    return self;
}

@end

File: android/app/src/main/java/com/teachbyfriends/TimelineView.java
package com.teachbyfriends;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.view.View;

public class TimelineView extends View {
    private double offsetBetweenLines;
    private int countOfLines;
    private double activeLine;
    private int activeLineInt;
    private double nRatio;
    private double linesRatio;
    private int width;
    private float regularLineWidth = 1.5f;
    private float mainLineWidth = 3f;
    private Paint regularLinePaint = new Paint();
    private Paint mainLinePaint = new Paint();
    private Paint mainLineTextPaint = new Paint();

    public double tPosition;
    public int countWordsOnScreen;
    public String[] timestamps;
    public double minAlpha;
    public double minWidthRatio;

    public TimelineView(Context context) {
        super(context);
        tPosition = 0.0;
        countOfLines = 0;
        countWordsOnScreen = 11;
        minAlpha = 0.1;
        minWidthRatio = 0.3;
        setNRatios();
        regularLinePaint.setColor(Color.BLACK);
        regularLinePaint.setStyle(Paint.Style.STROKE);
        regularLinePaint.setStrokeWidth(regularLineWidth);
        mainLinePaint.setColor(Color.BLACK);
        mainLinePaint.setStyle(Paint.Style.STROKE);
        mainLinePaint.setStrokeWidth(mainLineWidth);
        mainLineTextPaint.setColor(Color.BLACK);
        mainLineTextPaint.setTextAlign(Paint.Align.CENTER);
    }

    public void setTPosition(double value) {
        tPosition = value;
        invalidate();
        requestLayout();
    }

    public void setCountWordsOnScreen(int value) {
        if (value % 2 == 0) {
            countWordsOnScreen = value;
        } else {
            countWordsOnScreen = value + 1;
        }
        setNRatios();
        invalidate();
        requestLayout();
    }

    public void setTimestamps(String[] values) {
        timestamps = values;
        countOfLines = values.length;
        invalidate();
        requestLayout();
    }

    public void setMinAlpha(double value) {
        minAlpha = value;
        invalidate();
        requestLayout();
    }

    public void setMinWidthRatio(double value) {
        minWidthRatio = value;
        invalidate();
        requestLayout();
    }

    public void setNRatios() {
        int n = countWordsOnScreen / 2;
        nRatio = 0.55 + (float)n * 0.05;
        linesRatio = 8.0 - n;
        if (linesRatio < 0.0) {
            linesRatio = 3.0;
        }
    }

    @Override
    public void onSizeChanged(int w, int h, int oldW, int oldH) {
        width = w;
        offsetBetweenLines = (h - ((countOfLines - 1) * regularLineWidth) - mainLineWidth) / countOfLines;
        mainLineTextPaint.setTextSize((float)offsetBetweenLines);
        super.onSizeChanged(w, h, oldW, oldH);
    }

    @Override
    public void onDraw(Canvas canvas) {
        System.out.println(countOfLines);
        // TODO: set white color as background
        activeLine = tPosition / offsetBetweenLines;
        activeLineInt = (int) activeLine;
        for (int i = 0; i < countOfLines - 1; i++) {
            drawLine(i, canvas);
        }
    }

    private void drawLine(int lineNumber, Canvas canvas) {
        if (isMainLine(lineNumber)) {
            drawMainLine(lineNumber, canvas);
        } else {
            drawRegularLine(lineNumber, canvas);
        }
    }

    private void drawMainLine(int lineNumber, Canvas canvas) {
        float up = (float)(lineNumber * offsetBetweenLines);
        float width = (float)getLineWidth(lineNumber);
        // TODO: set white color behind text
        Rect textRect = new Rect();
        String positionAsString = String.valueOf(tPosition);
        mainLineTextPaint.getTextBounds(positionAsString, 0, positionAsString.length(), textRect);
        canvas.drawText(positionAsString, width / 2 - 1, up + textRect.height() / 2, mainLineTextPaint);
        // draw a lines from left and right sides over text
        canvas.drawLine(0, up, width / 2 - textRect.width() / 2 - 2, up, mainLinePaint);
        canvas.drawLine(width / 2 + textRect.width() / 2  + 2, up, width, up, mainLinePaint);
    }

    private void drawRegularLine(int lineNumber, Canvas canvas) {
        float up = (float)(lineNumber * offsetBetweenLines);
        float width = (float)getLineWidth(lineNumber);
        int lineAlpha = (int)(getLineAlpha(lineNumber)*100);
        regularLinePaint.setAlpha(lineAlpha);
        canvas.drawLine(0, up, width, up, regularLinePaint);
    }

    private boolean isMainLine(int currentLine) {
        double fraction = getActiveFraction(activeLine);
        int minActiveLine = activeLineInt;
        int maxActiveLine = minActiveLine + 1;
        if (fraction < 0.5 && currentLine == minActiveLine) {
            return true;
        } else if (fraction >= 0.5 && currentLine == maxActiveLine) { // fix > to >= in ios
            return true;
        }
        return false;
    }

    private double getActiveFraction(double activeLine) {
        return activeLine - (int)activeLine;
    }

    private double getLineAlpha(int lineNumber) {
        return minAlpha + (1 - ((double)Math.abs(lineNumber - activeLineInt)/countOfLines))*(1 - minAlpha);
    }

    private double getLineWidth(int lineNumber) {
        double additionalFactor = getAdditionalWidthRatio(lineNumber);
        double betweenBiggestFactor = additionalFactor * getBetweenBiggestWidthRatio(lineNumber);
        return (minWidthRatio +
                additionalFactor * (nRatio - minWidthRatio) +
                betweenBiggestFactor * (1 - nRatio)) * width;
    }

    private double getAdditionalWidthRatio(int lineNumber) {
        int linesToActive = getCountOfLinesToActive(lineNumber);
        return (1.0 - Math.tanh(linesToActive * linesRatio / 10));
    }

    private double getBetweenBiggestWidthRatio(int lineNumber) {
        double fraction = getActiveFraction(activeLine);
        if (lineNumber <= activeLineInt || fraction == 0.0) {
            return (1.0 - fraction);
        } else {
            return fraction;
        }
    }

    private int getCountOfLinesToActive(int lineNumber) {
        double fraction = getActiveFraction(activeLine);
        int minActiveLine = activeLineInt;
        int maxActiveLine = minActiveLine + 1;
        if (lineNumber <= activeLineInt || fraction == 0.0) {
            return Math.abs(lineNumber - minActiveLine);
        } else {
            return Math.abs(lineNumber - maxActiveLine);
        }
    }
}
